File: 2/Examples/VB/Reflection.Emit/HelloWorld.vb
''@ example:
''@ emit Emit
Imports System
Imports NUnit.Framework
Imports BLToolkit.Reflection.Emit

Namespace Examples.Reflection.Emit

    <TestFixture()> _
    Public Class HelloWorld

        Public Interface IHello
            Sub SayHello()
        End Interface

        <Test()> _
        Sub Test()
            Dim assemblyHelper As AssemblyBuilderHelper = New AssemblyBuilderHelper("HelloWorld.dll")
            Dim typeHelper As TypeBuilderHelper = assemblyHelper.DefineType("Hello", GetType(Object), GetType(IHello))
            Dim methodHelper As MethodBuilderHelper = typeHelper.DefineMethod(GetType(IHello).GetMethod("SayHello"))
            Dim emit As EmitHelper = methodHelper.Emitter

            Dim params() As Type = Array.CreateInstance(GetType(Type), 1)
            params(0) = GetType(String)

            emit.ldstr("Hello, World!")
            emit.call(GetType(Console).GetMethod("WriteLine", params))
            emit.ret()

            Dim type As Type = typeHelper.Create()

            Dim hello As IHello = Activator.CreateInstance(type)

            hello.SayHello()
        End Sub

    End Class

End Namespace

File: 2/Examples/CS/Reflection.Emit/HelloWorld.cs
//@ example:
//@ emit Emit
using System;

using NUnit.Framework;

using BLToolkit.Reflection;
using BLToolkit.Reflection.Emit;

namespace Examples.Reflection.Emit
{
	[TestFixture]
	public class HelloWorld
	{
		public interface IHello
		{
			void SayHello(string toWhom);
		}

		[Test]
		public void Test()
		{
			EmitHelper emit = new AssemblyBuilderHelper("HelloWorld.dll")
				.DefineType  ("Hello", typeof(object), typeof(IHello))
				.DefineMethod(typeof(IHello).GetMethod("SayHello"))
				.Emitter;

			/*[b]*/emit
				// string.Format("Hello, {0}!", toWhom)
				//
				.ldstr   ("Hello, {0}!")
				.ldarg_1
				.call    (typeof(string), "Format", typeof(string), typeof(object))

				// Console.WriteLine("Hello, World!");
				//
				.call    (typeof(Console), "WriteLine", typeof(string))
				.ret()
				;/*[/b]*/

			Type type = emit.Method.Type.Create();

			IHello hello = (IHello)TypeAccessor.CreateInstance(type);

			hello.SayHello("World");
		}
	}
}
